### PR TITLE
[MRG] Add LazyTensor for large scale OT

### PR DESCRIPTION
**Pull request by rflamary**
## Types of changes

This PR adds a very simple LazyTensor class that can implement large scale matrices and tensors that are computed only lazily when getting slices of the tensor.

This class will be very interesting when implementing large scale OT problems such as factored OT, low rank OT and lazy implementation of sinkhorn. The objective is to be very simple with a storage of all necessary data and a function that compute the values on sliced (without computing the full tensor).

Exemple of use

```python
import numpy as np
from ot.utils import LazyTensor, 

n1=100
n2=200
x1 = np.random.randn(n1,2)
x2 = np.random.randn(n2,2)

# i,j can be integers or slices, x1,x2 have to be passed as keyword arguments
def getitem(i,j, x1, x2):
    return np.dot(x1[i], x2[j].T)

# create a lazy tensor (giv data as arguments)
T = LazyTensor((n1,n2),getitem, x1=x1, x2=x2)

print(T.shape)
# (100, 200)
print(T)
# LazyTensor(shape=(100, 200),attributes=(x1,x2))

# get the full tensor (not lazy)
full_T = T[:]

# get one component
T11 = T[1,1]

# get one row
T1 = T[1]

# get one column with slices
Tsliced = T[::10,5]

# get the data as attributes
x1_0 = T.x1
x2_0 = T.x2

# create a LazyTensor from another one : exp(T)
T2 = ot.utils.LazyTensor(T.shape,lambda i,j,T: np.exp(T[i,j]) , T=T)

# compute sum
s = reduce_lazytensor(T, np.sum) 

# compute marginals
c = reduce_lazytensor(T, np.sum, axis=1) 
r = reduce_lazytensor(T, np.sum, axis=0, batch_size=50) 

```

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+    T[:]
+
+    # get one component
+    T[1, 1]
+
+    # get one row
+    T[1]
+
+    # get one column with slices
+    T[::10, 5]
```
Comment (agramfort) [inline test/test_utils.py:441]: can you add that the outputs are correct?

File: test/test_utils.py [hunk 2]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+
+
+def test_OTResult_LazyTensor(nx):
+
+    n1 = 100
+    n2 = 200
+
+    rng = np.random.RandomState(42)
+    a = rng.rand(n1)
+    a = a / a.sum()
```
Comment (agramfort) [inline test/test_utils.py:451]:
```suggestion
    a /= a.sum()
```

File: test/test_utils.py [hunk 3]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+    b = rng.rand(n2)
+    b = b / b.sum()
```
Comment (agramfort) [inline test/test_utils.py:453]:
```suggestion
    b /= b.sum()
```

File: test/test_utils.py [hunk 4]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+
+
+def test_LazyTensor_reduce(nx):
+
+    n1 = 100
+    n2 = 200
+
+    rng = np.random.RandomState(42)
+    a = rng.rand(n1)
+    a = a / a.sum()
```
Comment (agramfort) [inline test/test_utils.py:476]: /=

File: test/test_utils.py [hunk 5]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+    b = rng.rand(n2)
+    b = b / b.sum()
```
Comment (agramfort) [inline test/test_utils.py:478]: /=

File: test/test_utils.py [hunk 6]
```diff
@@ -401,3 +402,111 @@ def test_get_coordinate_circle():
...
+
+    a, b = nx.from_numpy(a, b)
+
+    def getitem(i, j, a, b):
+        return a[i, None] * b[None, j]
+
+    # create a lazy tensor
+    T = ot.utils.LazyTensor((n1, n2), getitem, a=a, b=b)
```
Comment (agramfort) [inline test/test_utils.py:486]: I would create fixture for T to avoid duplicating T creation in all your tests

File: ot/utils.py [hunk 7]
```diff
@@ -492,6 +492,79 @@ def get_coordinate_circle(x):
...
 
+def reduce_lazytensor(a, fun, axis=None, nx=None, batch_size=None):
+    """ Reduce a LazyTensor along an axis with function fun using batches.
+
+    When axis=None, reduce the LazyTensor to a scalar as a sum of fun over
+    batches taken along dim.
+
+    Parameters
+    ----------
+
```
Comment (agramfort) [inline ot/utils.py:503]:
```suggestion
```

File: ot/utils.py [hunk 8]
```diff
@@ -492,6 +492,79 @@ def get_coordinate_circle(x):
...
+        If 0 or 1 reduce the LazyTensor to a vector/matrix as a sum of fun over
+        batches taken along axis.
+    nx : Backend, optional
+        Backend to use for the reduction
+    batch_size : int, optional
+        Size of the batches to use for the reduction (default=100)
+
+    Returns
+    -------
+
```
Comment (agramfort) [inline ot/utils.py:520]:
```suggestion
```

File: ot/utils.py [hunk 9]
```diff
@@ -492,6 +492,79 @@ def get_coordinate_circle(x):
...
+    res : array-like
+        Result of the reduction
+
+    """
+
+    if nx is None:
+        nx = get_backend(a[0])
+
+    if batch_size is None:
+        batch_size = 100
```
Comment (agramfort) [inline ot/utils.py:530]: make 100 the default in signature

File: ot/utils.py [hunk 10]
```diff
@@ -492,6 +492,79 @@ def get_coordinate_circle(x):
...
+
+    When axis=None, reduce the LazyTensor to a scalar as a sum of fun over
+    batches taken along dim.
+
+    Parameters
+    ----------
+
+    a : LazyTensor
+        LazyTensor to reduce
+    fun : callable
```
Comment (agramfort) [inline ot/utils.py:506]: I would call this func rather than fun

File: ot/utils.py [hunk 11]
```diff
@@ -492,6 +492,79 @@ def get_coordinate_circle(x):
     return x_t
 
 
+def reduce_lazytensor(a, fun, axis=None, nx=None, batch_size=None):
+    """ Reduce a LazyTensor along an axis with function fun using batches.
```
Comment (agramfort) [inline ot/utils.py:496]: the code it's just for 2d arrays/tensors. Do you confirm?

Comment (rflamary) [reply ot/utils.py:496]: updated the doc, t should work for any shape but reuction is imllemented only for global and axis 1 and 2

**Pull request merged**